### ISSUE TITLE
Loss of subject reduction with Setω

### ISSUE THREAD

**Issue by L-TChen** [agda/agda#5810]
Subject reduction is lost with `Setω`.  Consider 
```agda
open import Agda.Primitive
open import Agda.Builtin.Unit

bad : Setω
bad = (λ (ℓ : ⊤ → Level) → ∀ t → Set (ℓ t)) (λ _ → lzero)
```
whose type is `Setω` but after normalisation it becomes
```agda
bad = (t : ⊤) → Set
```
whose type is now `Set₁`. Indeed, if we fill the hole with the normalised expression by `C-c C-m`, Agda happily produces the second expression and complains about the term it just produced ... Any idea?

**Comment by nad** [comment]
The only theoretical work in this area (first-class level types and `Setω`) that I am aware of is @AndrasKovacs' [Generalized Universe Hierarchies and First-Class Universe Levels](https://doi.org/10.48550/arXiv.2103.00223). However, that paper does not seem to address subject reduction.

**Comment by andreasabel** [comment]
In type theories with subtyping, reduction may produce something of a _more precise_ type.  So, with `--cumulativity`, this would be perfectly fine.
Agda's double checker uses cumulativity _always_, independent of flag `--cumulativity`, so this subject reduction "problem" isn't found by the double checker:
```agda
{-# OPTIONS --no-cumulativity #-}
{-# OPTIONS --double-check #-}
{-# OPTIONS -v tc.check.internal:20 #-}

open import Agda.Primitive

record ⊤ : Set where

bad : Setω
bad = (λ (ℓ : ⊤ → Level) → ∀ t → Set (ℓ t)) (λ _ → lzero)

-- checking internal  (t : ⊤) → Set lzero : Setω
```

**Comment by jespercockx** [comment]
Perhaps it would make more sense to only assign type `Setω` to a pi type `(x : A) → Set e` if the variable `x` occurs *strongly rigid* in the expression `e`. This breaks backwards compatibility, but that's probably a price worth paying to preserve subject reduction.

**Comment by jespercockx** [comment]
Actually the occurrence of `t` in `ℓ t` is counted as strongly rigid, so the solution I suggested above does not work.

**Comment by jespercockx** [comment]
I have implemented an experimental fix for this issue, which you can find at https://github.com/jespercockx/agda/tree/Issue5810. However the builtin cubical modules fail to compile with it. Specifically, I get an error on the type of `primComp`:

```agda
  primComp : ∀ {ℓ} (A : (i : I) → Set (ℓ i)) {φ : I} (u : ∀ i → Partial φ (A i)) (a : A i0) → A i1
```

The problem here is in the type of `u`, which is `∀ i → Partial φ (A i)`. The type of of `Partial φ (A i)` is `SSet (ℓ i)`, and since this depends on `i`, the type `∀ i → Partial φ (A i)` gets assigned type `SSetω`. However this exhibits exactly the problem we are talking about in this issue: if `ℓ` is instantiated with a constant function, say `λ _ → lzero`, the type of `Partial φ (A i)` becomes `SSet lzero`, and hence the `∀ i → Partial φ (A i)` now has type `SSet lzero` too. So already our builtin modules contain expressions that do not satisfy subject reduction in this manner!

I'm summoning the cubical overlords @Saizan @mortberg @plt-amy so hopefully they can give me an idea what the sort of `∀ i → Partial φ (A i)` *should* be here.

**Comment by plt-amy** [comment]
Ah, we're being bitten in the behind by comp/transp taking a line of levels. Here's what @Saizan had to say on this when it last came up: https://github.com/agda/agda/issues/3722#issuecomment-556000316. I think the right move is:

> Now, the sort of B should not actually depend on x when the sort of (x : A) -> B is not set-omega, however i did not feel it was safe to make use of this invariant at the time. Perhaps I should have. (Then one could strengthen the sort of B).

We can't transport/compose in Setω anyway, so it wouldn't be a big loss to change primComp/primTransp to take `Level` rather than `I → Level`.

**Comment by jespercockx** [comment]
So to experiment I disabled my check when `--cubical` is enabled and (while many tests then pass) there are a number of other test cases that then fail:

- [ ] Succeed/Issue930.agda contains the postulate `P : (a : ⊤) → Set (q a)` (though `q` is also a postulate so it cannot actually be instantiated). https://github.com/agda/agda/blob/d50b1e967818974b2af797c2641d2c8f9db1ede9/test/Succeed/Issue930.agda#L28
- [ ] Succeed/Issue2408.agda contains the postulate `F : .(a : A) → Set (l a)` (though `l` is also a postulate so it cannot actually be instantiated) https://github.com/agda/agda/blob/d50b1e967818974b2af797c2641d2c8f9db1ede9/test/Succeed/Issue2408.agda#L15
- [ ] Succeed/Issue2408Unused.agda contains the postulate `F : (a : ⊥) → Set (l u a)` (though `l u a` is blocked on the postulate `u` so it cannot actually be reduced). https://github.com/agda/agda/blob/d50b1e967818974b2af797c2641d2c8f9db1ede9/test/Succeed/Issue2408Unused.agda#L23
- [ ] Fail/Issue2408/LevelDependentOnIrrelevant contains the postulate `.(a : A) → Set (l a)` https://github.com/agda/agda/blob/d50b1e967818974b2af797c2641d2c8f9db1ede9/test/Fail/Issue2408/LevelDependentOnIrrelevant.agda#L14
- [ ] Succeed/Cumulativity.agda has a bunch of unsolved constraints.

In addition, many error messages get more cluttered with extra "Has PTS rule: ..." constraints (Issue399.agda, Issue2344.agda, Issue2420.agda, Issue2834.agda, Issue2927.agda, Issue3044.agda, Issue4401.agda, ...) Personally I am more worried about these because they will make Agda's constraints (even) harder to read. Since these PTS constraints always arise from unsolved metavariables, perhaps we could hide them from the constraint output by default (and allow toggling them on with a flag perhaps).

I am also trying my fix on the std-lib, and I run into problems with the `Nary` constructions, specifically `Sets` in `Function.Nary.Nondependent.Base`:
```agda
Sets : ∀ n (ls : Levels n) → Set (Level.suc (⨆ n ls))
```
This shows a more serious limitation of my current attempt: the level `Level.suc (⨆ n ls)` is blocked on `n` first and not on `ls`, so `(ls : Levels n) → Set (Level.suc (⨆ n ls))` is not accepted by my check. 

It feels to me that this might require some more discussion and careful thought before we can have a fix that is both sufficient and does not break a bunch of sensible existing code that plays around with universe levels. Because of this and since this is not a regression, I am bumping this to 2.6.4.

**Comment by andreasabel** [comment]
Setomega is really a truncation of what would be naturally ordinal levels.  Say we have construction for limit levels 
```
sup : (A : Set) (f : A -> Level) -> Level
```
then `(x : A) -> Set (f x)` does not live a `Setomega` but simply in `Set (sup A \ x -> lsuc (f x))`.  Once `f x = lzero` is known, this collapses to `Set (sup A \ _ -> lsuc zero) = Set1`.

We currently _round up_ every limit on levels to omega and assign universe `Setomega` as safe choice.  Of course, this loses information which might be regained by reduction.  
So, the principled solution for the OP is to have cumulativity for `Setomega`.  Then the gained information can be discarded, as wished, by the subsumption rule, and we have "subject reduction" back.

**Comment by nad** [comment]
If cumulativity is added for `Setω`, then I think `Setω` should be hidden behind a flag. (I think that `Setω` should be hidden behind a flag anyway, but I think the case for that is stronger with cumulativity.)

**Comment by L-TChen** [comment]
I’d second @andreasabel’s suggestion. The text of this issue might be misleading as I didn’t know what to expect at first, so I hinted that subject reduction should be restored. 

[ On the other hand, not directly relevant to this issue, what is first-class universe level anyway? Can we learn something from @AndrasKovacs‘s work and lay the groundwork of what Agda is using precisely? ]

**Comment by andreasabel** [comment]
> If cumulativity is added for `Setω`, then I think `Setω` should be hidden behind a flag. (I think that `Setω` should be hidden behind a flag anyway, but I think the case for that is stronger with cumulativity.)

Could you elaborate why?

**Comment by nad** [comment]
Lots of code breaks if `--cumulativity` is enabled, presumably because it is harder to infer levels.

Furthermore the current implementation of cumulativity is experimental, it is hidden behind a flag, and it is not compatible with `--safe`.

**Comment by andreasabel** [comment]
Yes, but we could make Setomega cumulative and leave the other universes alone.  So, we do not need to enable `--cumulativity` in general.

**Comment by jespercockx** [comment]
Having subtyping `Set i <: Setomega` is an interesting idea, fixing the problem by making Agda more liberal rather than more restrictive. I can see both advantages and disadvantages of this approach:
- The advantage is that it is backwards compatible and arguably more intuitive than some kind of syntactic restriction
- The disadvantage is that it makes Agda's metatheory more complicated, and also (re)introduces subtyping to Agda which has been problematic wrt constraint solving.

**Comment by nad** [comment]
I don't want to have any subtyping enabled by default (unless perhaps if someone finds a good way to combine subtyping with Agda's treatment of meta-variables).

**Comment by nad** [comment]
I just finished [adding](https://github.com/graded-type-theory/graded-type-theory/pull/89) infinite universe levels (ω, ω + 1, and so on) to the graded type theory formalisation. In that formalisation the type corresponding to `(l : Level) → Set l` does not have a type: `(x : A) → B` has type `Set l` if `A` and `B` both have type `Set l`: note that `l` can not refer to `x`. The formalisation currently requires `A` and `B` to have the same universe level: one can use lifting to encode more general types. One can also use lifting to encode something akin to `(l : Level) → Set l`.

One way to fix the issue with subject reduction might be to use a typing rule like the following one for Π: `(A : Set l₁) → (B : A → Set l₂) → Set (l₁ ⊔ l₂)` (where `l₁` and `l₂` can be infinite). Note that the argument of `B` is not in scope in `l₂`. One can then encode things like `(l : Level) → Set l`:
```agda
open import Agda.Primitive

private variable
  a b : Level

Π : (A : Set a) → (B : A → Setω) → Setω
Π A B = (x : A) → B x

record Liftω (A : Set a) : Setω where
  constructor lift
  field
    lower : A

_ : Setω
_ = Π Level (λ l → Liftω (Set l))
```
Note that `Liftω` stays in the type even if its argument is reduced:
```agda
_ : Setω
_ = (λ (ℓ : ⊤ → Level) → (x : ⊤) → Liftω (Set (ℓ x))) (λ _ → lzero)
```

I wonder how much code would break if we took this approach.

**Comment by nad** [comment]
> I wonder how much code would break if we took this approach.

One couldn't even give the type of a universe-polymorphic data type like `List` in the usual way, so the answer is presumably "a lot".

**Comment by nad** [comment]
One could imagine elaborating surface syntax into something with annotations as discussed above. However, it should be possible to print the result of evaluating something.

Internally Agda's Π-types are annotated with levels (sorts). With a suitable design that involved such annotations as part of the surface syntax one could perhaps avoid issues with subject reduction. Can we come up with an annotation scheme that would in most cases be invisible to users?

Let me restrict attention to sorts of the form `Set l`, where `l` can be finite or infinite. The notation `(@(level l) x : A) → B` could mean that the level of `B` is at most `l`, and that `l` is used to compute the sort of the Π-type:

* We would have `(@(level l₂) x : A) → B : Set (l₁ ⊔ l₂)` if `A : Set l₁`, `B : Set l₂′` and `l₂′ ≤ l₂`. Here `l₂′` can mention `x`, but `x` is not in scope in `l₂`.

* If the annotation is omitted, then it is interpreted as `l₂′` if `x` is not free in `l₂′` *after normalisation*, and otherwise as `ω`. In the presence of meta-variables the annotation could perhaps not be computed, then its computation should be postponed. (One obvious question is how expensive the normalisation would be. I suspect that arguments of `Set` are rarely very complicated.)

* When printing the annotation is included only if necessary.

Let's consider the example from the OP, rewritten slightly:
```agda
bad : Setω
bad = (λ (ℓ : ⊤ → Level) → (t : ⊤) → Set (ℓ t)) (λ _ → lzero)
```
Here there is no need to annotate `(t : ⊤) → Set (ℓ t)`: the type of `Set (ℓ t)` is `Set (lsuc (ℓ t))`, and `t` is free in `lsuc (ℓ t)`, so `(t : ⊤) → Set (ℓ t)` means `(@(level ω) t : ⊤) → Set (ℓ t)`. After reduction we get `(@(level ω) t : ⊤) → Set₀`: now the annotation cannot be omitted, because `Set₀ : Set₁`, and `t` is not free in `Set₁`, so the inferred annotation would be `1`, not `ω`.

With this approach the type of `List` is fine without annotations: `List : {a : Level} (A : Set a) → Set a`. The inferred annotations are correct: `List : {@(level ω) a : Level} (@(level (lsuc a)) A : Set a) → Set a`. The type of `List` would have the type `Setω`.

What do you think? Could something like this be made to work?

**Comment by AndrasKovacs** [comment]
@nad I thought about this some time ago and my suggested design would be kinda similar to yours.

If we only had Setω and finite level polymorphism, the best design would be to follow the level normalization paper and have a Bool flag on Pi that distinguishes dependently sized and non-dependently sized variants. The  
variants are definitionally distinct. There could be explicit user syntax but by default the flag would be inferred by elaboration. But we have infinite levels, so this setup isn't quite enough. 

So I suggest:

- Have non-dependently and dependently sized Pi as I mentioned. 
- The level of a dependently sized Pi is the following. We have `A : Set i`, `a : A ⊢ j InfLevel` and `a : A ⊢ B : Set j`, where `InfLevel` includes finite and infinite levels. Then the level of `Pi i A j B`  is `i ⊔ ω ⊔ j[a]` where `a : A` is a fresh variable, but the result is considered blocked if `a` is free in it. 

Now, as far as I see, the only way `a` can ever be free in the output is through unsolved metavariables, becuse Agda only supports finite dependent levels, so whatever level dependency we have would be absorbed by `ω ⊔`. Hence, we're only postponing computed sorts because of metavariables/constraints, which is business as usual. 

This seems similar to @nad's suggestion but we only have a Bool worth of Pi annotation, and the level computation is simpler.

**Comment by nad** [comment]
> Then the level of `Pi i A j B`  is `i ⊔ ω ⊔ j[a]` where `a : A` is a fresh variable, but the result is considered blocked if `a` is free in it.

Unless I've misunderstood something this would mean that the level is always at least ω.

**Comment by AndrasKovacs** [comment]
That's correct. If we want to use functions at finite levels, we need to use the other Pi type which does not allow dependent codomain level. I believe in current Agda we also get at least Set_omega for dependently sized function types.

**Comment by jespercockx** [comment]
I am in support of the proposal by @AndrasKovacs : basically we need to have a core type theory with two separate pi types: one with level dependencies but only small sorts (which always ends up in `Setω`), and one without level dependencies that also works with large sorts (which lives in the `funSort` of the domain and codomain sorts). For backwards compatibility, the elaborator would have to infer which of these is meant by normalizing the sort of the codomain and checking for any dependency, blocking type-checking of the whole expression if the answer is inconclusive because of metavariables. But we could also allow an explicit annotation (e.g. `@dependent-level`) both for letting the user disambiguate and for pretty-printing. Note that annotating the actual level itself is not necessary, since the only place where the annotation would be used is if the overall sort of the pi type is `Setω`.

Perhaps I could work on implementing this feature during the upcoming Agda meeting, though I won't be present during the full week.

**Comment by AndrasKovacs** [comment]
@jespercockx in my proposal the dependently sized Pi can be larger than Set_omega and the domain and codomain can be both large. That would be also important for compatibility, e.g. in current Agda the following is accepted:

```agda
open import Level

record Foo : Setω₂ where
  field
    fst : Level
    snd : Setω
open Foo

Ty : Setω₂
Ty = (x : Foo) → Set (fst x)
```

**Comment by jespercockx** [comment]
Well if the codomain is large then it cannot be dependent, so you don't need a level-dependent version of that. But I agree that your example shows that the *domain* can be large while the codomain is still dependent. But actually if the domain is large it does not matter whether the codomain is dependent, since the pi type overall will be large either way. So there is no problem with subject reduction in this case, which is why I didn't consider it.

**Comment by AndrasKovacs** [comment]
OK I see that large codomains can't be dependent. Still, I prefer if one Pi type variant can be dependent and the other can't be, in which case the Pi types with large domain and dependent codomain have to be in the dependent variant.

**Comment by jespercockx** [comment]
I'm working on the new proposed solution at https://github.com/jespercockx/agda/tree/issue5810-try2. Currently it adds an annotation `@leveldep` (suggestions for better names welcome) that creates a separate function space of level-dependent pi types that are always large, and disallows level-dependency in normal pi types.

To ensure backwards compatibility, I'm trying to add some heuristics for inferring these annotations, but this proves to be rather difficult to do well. One example where the current simple heuristic behaves strangely is the following:

```agda
open import Agda.Primitive

test1 = (l1 l2 : Level) → Set l1 → Set l2
test2 = (l1 : Level) → (l2 : Level) → Set l1 → Set l2
```

For `test1` both level parameters get a `@leveldep` annotation, but in the second example only `l2` gets one. This is also valid because the binding of `l1` is not actually level dependent (as the codomain is in `Setω`) but it is surprising that the levels would get different annotations based on whether they are bound together or separately.

Another tricky case is how to infer these annotations for module parameters: it might be the case that a module just has a universe level as a parameter, and whether it should get a `@leveldep` annotation depends on whether any of the definitions in the module have a sort that is dependent on the level or not. But that means we could only make the decision *after* checking the whole module, which seems dangerous.

**Comment by nad** [comment]
> Currently it adds an annotation `@leveldep` (suggestions for better names welcome) that creates a separate function space of level-dependent pi types that are always large, and disallows level-dependency in normal pi types.

I think there should be annotations for both variants.

> Another tricky case is how to infer these annotations for module parameters: it might be the case that a module just has a universe level as a parameter, and whether it should get a `@leveldep` annotation depends on whether any of the definitions in the module have a sort that is dependent on the level or not. But that means we could only make the decision _after_ checking the whole module, which seems dangerous.

I think it is unfortunate if inlining of a module telescope could change the meaning of the type. Could the inference procedure ensure that this is never the case? With two annotations the user can always decide what kind of Π-type to use, irrespective of what the inference procedure does.

**Comment by jespercockx** [comment]
Sure, adding an annotation for non-level dependent pi types is not a problem. Do you have a suggestion for a name that's better than `@nonleveldep`?

> I think it is unfortunate if inlining of a module telescope could change the meaning of the type. Could the inference procedure ensure that this is never the case?

In theory we could postpone the decision until after the whole module has been checked. However, this would involve traversing all the definitions in the module in order to modify their types, which I'd prefer to avoid if possible.

**Comment by nad** [comment]
> In theory we could postpone the decision until after the whole module has been checked. However, this would involve traversing all the definitions in the module in order to modify their types, which I'd prefer to avoid if possible.

It might still be the case that you get different types:
```agda
module _ (a : Level) where

  b : Level
  b = a

  A : Set (lsuc a)
  A = Set a
```
Here the Π for `a` would be a level-dependent one, so both `b` and `A` would get such types. However, if you inline the telescope, then you only get that kind of Π for `A`, not for `b`.
